src/components/map/EventPopup.js

```diff
@@ -0,0 +1,216 @@
+
+    if (Array.isArray(programStageDataElements)) {
+        displayElements = programStageDataElements
+            .filter(d => d.displayInReports)
+            .map(d => d.dataElement);
+
+        for (let d of displayElements) {
+            await loadOptionSet(d);
+        }
+
+        if (eventCoordinateField) {
+            const coordElement = programStageDataElements.find(
+                d => d.dataElement.id === eventCoordinateField
+            );
+
+            if (coordElement) {
+                eventCoordinateFieldName = coordElement.dataElement.name;
+            }
+        }
+    }
+
+    return { displayElements, eventCoordinateFieldName };
+};
+
+// Loads event data for the selected feature
+const loadEventData = async feature =>
+    feature ? apiFetch(`/events/${feature.id}`) : null;
+
+// Returns table rows for all display elements
+const getDataRows = (displayElements, dataValues, styleDataItem, value) => {
+    const dataRows = [];
+
+    // Include data element used for styling if not included below
+    if (
+        styleDataItem &&
+        !displayElements.find(d => d.id === styleDataItem.id)
+    ) {
+        dataRows.push(
+            <tr key={styleDataItem.id}>
+                <th>{styleDataItem.name}</th>
+                <td>{hasValue(value) ? value : i18n.t('Not set')}</td>
+            </tr>
+        );
+    }
+
+    // Include rows for each displayInReport data elements
+    displayElements.forEach(({ id, name, valueType, options }) => {
+        const { value } = dataValues.find(d => d.dataElement === id) || {};
+        let formattedValue = value;
+
+        if (valueType === 'COORDINATE' && value) {
+            formattedValue = formatCoordinate(value);
+        } else if (options) {
+            formattedValue = options[value];
+        } else if (!hasValue(value)) {
+            formattedValue = i18n.t('Not set');
+        }
+
+        dataRows.push(
+            <tr key={id}>
+                <th>{name}</th>
+                <td>{formattedValue}</td>
+            </tr>
+        );
+    });
+
+    if (dataRows.length) {
+        dataRows.push(<tr key="divider" style={{ height: 5 }} />);
+    }
+
+    return dataRows;
+};
+
+// Will display a popup for an event feature
+const EventPopup = props => {
+    const {
+        coordinates,
+        feature,
+        programStage,
+        eventCoordinateField,
+        styleDataItem,
+        onClose,
+    } = props;
+
+    const [eventData, setEventData] = useState();
+    const [dataElements, setDataElements] = useState({});
+    const { displayElements = [], eventCoordinateFieldName } = dataElements;
+
+    // Load data elements every time programStage is changed
+    useEffect(() => {
+        loadDataElements(programStage, eventCoordinateField).then(
+            setDataElements
+        );
+    }, [programStage]);
+
+    // Load event data every time a new feature is clicked
+    useEffect(() => {
+        loadEventData(feature).then(setEventData);
```

Comment:
To be pedantic - there is a race condition here.  If `loadEventData` is called and then the component unmounts before the promise resolves, the `then()` clause will still execute and try to set state on an unmounted component (an error/warning in React).  To avoid this we need to track an **aborted** state, something like this should work:

```js
useEffect(() => {
  let aborted = false;
  loadEventData(feature).then(data => {
    if (!aborted) {
      setEventData(data)
    }
  })
  return () => {
    aborted = true;
  }
}, [feature])
```

Reply:
Fixed.